### PR TITLE
Task-50876: Implement delete news target from Manage news publication targets page

### PR DESCRIPTION
**Pull request by hassenelkamel-exo**
After this change, we implement delete target and undo deleting services.

### CODE REVIEW

File: services/src/test/java/org/exoplatform/news/NewsTargetingImplTest.java [hunk 1]
```diff
@@ -38,7 +40,7 @@
   
   @Mock
   IdentityManager identityManager;
-
+
```
Comment (azayati) [inline services/src/test/java/org/exoplatform/news/NewsTargetingImplTest.java]: Implement UT for deleteTargetByName method

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 2]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+              RequestLifeCycle.end();
+            }
+          }
+        }, delay, TimeUnit.SECONDS);
+      } else {
+        newsTargetToDeleteQueue.remove(targetName);
+        newsTargetingService.deleteTargetByName(targetName, currentIdentity);
+      }
+      return Response.ok().build();
+    } catch (Exception e) {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java:163]: you need to catch IllegalAccessException and return Response.Status.UNAUTHORIZED, see NewsRestResourcesV1 methods

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 3]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+      LOG.warn("Error when deleting the news target");
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
      LOG.error("Error when deleting the news target with name " + targetName, e);
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 4]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+        newsTargetToDeleteQueue.put(targetName, currentIdentity.getUserId());
+        scheduledExecutor.schedule(() -> {
+          if (newsTargetToDeleteQueue.containsKey(targetName)) {
+            ExoContainerContext.setCurrentContainer(container);
+            RequestLifeCycle.begin(container);
+            try {
+              newsTargetToDeleteQueue.remove(targetName);
+              newsTargetingService.deleteTargetByName(targetName, currentIdentity);
+            }  catch (Exception e) {
+              LOG.warn("Error when deleting the news target with name " + targetName, e);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
              LOG.error("Error when deleting the news target with name " + targetName, e);
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 5]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+  @DELETE
+  @Path("{targetName}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @RolesAllowed("users")
+  @ApiOperation(value = "Delete target news", httpMethod = "DELETE", response = Response.class, notes = "This deletes the target news", consumes = "application/json")
+  @ApiResponses(value = { @ApiResponse(code = 200, message = "News target deleted"),
+          @ApiResponse(code = 400, message = "Invalid query input"),
+          @ApiResponse(code = 401, message = "User not authorized to delete the news target"),
+          @ApiResponse(code = 500, message = "Internal server error") })
+  public Response deleteTargetByName(@Context HttpServletRequest request,
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
  public Response deleteTarget(@Context HttpServletRequest request,
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 6]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+  )
+  @ApiResponses(
+          value = {
+                  @ApiResponse(code = HTTPStatus.NO_CONTENT, message = "Request fulfilled"),
+                  @ApiResponse(code = HTTPStatus.BAD_REQUEST, message = "Invalid query input"),
+                  @ApiResponse(code = HTTPStatus.FORBIDDEN, message = "Forbidden operation"),
+                  @ApiResponse(code = HTTPStatus.UNAUTHORIZED, message = "Unauthorized operation"),
+                  @ApiResponse(code = HTTPStatus.INTERNAL_ERROR, message = "Internal server error"), }
+  )
+  public Response undoDeleteNewsTarget(@Context HttpServletRequest request,
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
  public Response undoDeleteTarget(@Context HttpServletRequest request,
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 7]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+        newsTargetingService.deleteTargetByName(targetName, currentIdentity);
+      }
+      return Response.ok().build();
+    } catch (Exception e) {
+      LOG.warn("Error when deleting the news target");
+      return Response.serverError().entity(e.getMessage()).build();
+    }
+  }
+
+  @Path("{targetName}/undoDeleteNewsTarget")
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
  @Path("{targetName}/undoDelete")
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 8]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+  @POST
+  @RolesAllowed("users")
+  @ApiOperation(
+          value = "Undo deleting target news if not yet effectively deleted.",
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
          value = "Undo deleting news target if not yet effectively deleted.",
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 9]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
     }
   }
 
+  @DELETE
+  @Path("{targetName}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @RolesAllowed("users")
+  @ApiOperation(value = "Delete target news", httpMethod = "DELETE", response = Response.class, notes = "This deletes the target news", consumes = "application/json")
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
  @ApiOperation(value = "Delete news target", httpMethod = "DELETE", response = Response.class, notes = "This deletes the target news", consumes = "application/json")
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 10]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+  @POST
+  @RolesAllowed("users")
+  @ApiOperation(
+          value = "Undo deleting target news if not yet effectively deleted.",
+          httpMethod = "POST",
+          response = Response.class
+  )
+  @ApiResponses(
+          value = {
+                  @ApiResponse(code = HTTPStatus.NO_CONTENT, message = "Request fulfilled"),
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]: useless

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 11]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
     }
   }
 
+  @DELETE
+  @Path("{targetName}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @RolesAllowed("users")
+  @ApiOperation(value = "Delete target news", httpMethod = "DELETE", response = Response.class, notes = "This deletes the target news", consumes = "application/json")
+  @ApiResponses(value = { @ApiResponse(code = 200, message = "News target deleted"),
+          @ApiResponse(code = 400, message = "Invalid query input"),
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
          @ApiResponse(code = HTTPStatus.BAD_REQUEST, message = "Invalid query input"),
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 12]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+          @ApiResponse(code = 401, message = "User not authorized to delete the news target"),
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
          @ApiResponse(code = HTTPStatus.UNAUTHORIZED, message = "User not authorized to delete the news target"),
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 13]
```diff
@@ -79,4 +109,89 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+          @ApiResponse(code = 500, message = "Internal server error") })
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
          @ApiResponse(code = HTTPStatus.INTERNAL_ERROR, message = "Internal server error") })
```

**Review by azayati — CHANGES_REQUESTED** [review]
see review

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java (hunk 2 above)
Comment (hassenelkamel-exo) [reply services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java:163]: Exception 'java.lang.IllegalAccessException' is never thrown in the corresponding try block

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 14]
```diff
@@ -89,4 +118,86 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+      if (delay > 0) {
+        newsTargetToDeleteQueue.put(targetName, currentIdentity.getUserId());
+        scheduledExecutor.schedule(() -> {
+          if (newsTargetToDeleteQueue.containsKey(targetName)) {
+            ExoContainerContext.setCurrentContainer(container);
+            RequestLifeCycle.begin(container);
+            try {
+              newsTargetToDeleteQueue.remove(targetName);
+              newsTargetingService.deleteTargetByName(targetName, currentIdentity);
+            } catch (Exception e) {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java:151]: You should catch the illegalAccess exception here

**Review by azayati — CHANGES_REQUESTED** [review]
See review

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 15]
```diff
@@ -89,4 +118,91 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+        newsTargetToDeleteQueue.put(targetName, currentIdentity.getUserId());
+        scheduledExecutor.schedule(() -> {
+          if (newsTargetToDeleteQueue.containsKey(targetName)) {
+            ExoContainerContext.setCurrentContainer(container);
+            RequestLifeCycle.begin(container);
+            try {
+              newsTargetToDeleteQueue.remove(targetName);
+              newsTargetingService.deleteTargetByName(targetName, currentIdentity);
+            } catch (IllegalAccessException e) {
+              LOG.error("Error when deleting the news target with name " + targetName, e);
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
              LOG.warn("User '{}' is not autorized to delete the news target with name " + targetName, e);
              return Response.status(Response.Status.UNAUTHORIZED).entity(e.getMessage()).build();
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 16]
```diff
@@ -89,4 +118,91 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+      String originalModifierUser = newsTargetToDeleteQueue.get(targetName);
+      if (!originalModifierUser.equals(authenticatedUser)) {
+        LOG.warn("User {} attempts to cancel deletion of a news target deleted by user {}", authenticatedUser, originalModifierUser);
+        return Response.status(Response.Status.FORBIDDEN).build();
+      }
+      newsTargetToDeleteQueue.remove(targetName);
+      return Response.noContent().build();
+    } else {
+      return Response.status(Response.Status.BAD_REQUEST)
+                     .entity("Target News with name {} was already deleted or isn't planned to be deleted" + targetName)
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
                     .entity("News target with name {} was already deleted or isn't planned to be deleted" + targetName)
```

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 17]
```diff
@@ -89,4 +118,91 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
     }
   }
 
+  @DELETE
+  @Path("{targetName}")
+  @Produces(MediaType.APPLICATION_JSON)
+  @RolesAllowed("users")
+  @ApiOperation(value = "Delete news target", httpMethod = "DELETE", response = Response.class, notes = "This deletes the target news", consumes = "application/json")
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]:
```suggestion
  @ApiOperation(value = "Delete news target", httpMethod = "DELETE", response = Response.class, notes = "This deletes the news target", consumes = "application/json")
```

File: services/src/main/java/org/exoplatform/news/service/impl/NewsTargetingServiceImpl.java [hunk 18]
```diff
@@ -66,7 +66,10 @@ public NewsTargetingServiceImpl(MetadataService metadataService, IdentityManager
   }
   
   @Override
-  public void deleteTargetByName(String targetName) {
+  public void deleteTargetByName(String targetName, org.exoplatform.services.security.Identity  currentIdentity) throws IllegalAccessException {
+    if (currentIdentity != null && !NewsUtils.canPublishNews(currentIdentity)) {
+      throw new IllegalArgumentException("User " + currentIdentity.getUserId() + " not authorized to publish news");
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/service/impl/NewsTargetingServiceImpl.java]:
```suggestion
      throw new IllegalArgumentException("User " + currentIdentity.getUserId() + " not authorized to delete news target with name " + targetName");
```

**Review by azayati — CHANGES_REQUESTED** [review]
see review

File: services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java [hunk 19]
```diff
@@ -89,4 +118,91 @@ public Response getReferencedTargets(@Context HttpServletRequest request) {
...
+              RequestLifeCycle.end();
+            }
+          }
+        }, delay, TimeUnit.SECONDS);
+      } else {
+        newsTargetToDeleteQueue.remove(targetName);
+        newsTargetingService.deleteTargetByName(targetName, currentIdentity);
+      }
+      return Response.ok().build();
+    } catch (IllegalAccessException e) {
```
Comment (azayati) [inline services/src/main/java/org/exoplatform/news/rest/NewsTargetingRestResourcesV1.java]: not needed already catched above

**Review by azayati — CHANGES_REQUESTED** [review]
see review

**Review by azayati — APPROVED** [review]
LGTM

**Pull request merged**